libcextract/SymbolExternalizer.cpp

```diff
@@ -71,142 +87,161 @@ using namespace llvm;
+    SymbolUpdateStatus *sym = SE.getSymbolsUpdateStatus(decl->getName());
+    if (sym == nullptr)
+      return VISITOR_CONTINUE;
+
+    ExternalizationType type = sym->ExtType;
+    if (type == ExternalizationType::STRONG) {
+      if (!sym->Done) {
+        std::string sym_name = decl->getName().str();
+
+        /* If we found the first instance of the function we want to externalize,
+           then proceed to create and replace the function declaration node with
+           a variable declaration node of proper type.  */
+        std::string new_name = EXTERNALIZED_PREFIX + sym_name;
+        sym->new_name = new_name;
+        DeclaratorDecl *new_decl = SE.Create_Externalized_Var(decl, new_name);
+        SE.Log.push_back({.OldName = sym_name,
+                       .NewName = new_name,
+                       .Type = type});
+
+        /* Create a string with the new variable type and name.  */
+        std::string o;
+        llvm::raw_string_ostream outstr(o);
+
+        /*
+         * It won't be a problem to add the code below multiple times, since
+         * clang-extract will remove ifndefs for already defined macros
+         */
+        if (SE.Ibt) {
+          outstr << "#ifndef KLP_RELOC_SYMBOL_POS\n"
+                    "# define KLP_RELOC_SYMBOL_POS(LP_OBJ_NAME, SYM_OBJ_NAME, SYM_NAME, SYM_POS) \\\n"
+                    "   asm(\"\\\".klp.sym.rela.\" #LP_OBJ_NAME \".\" #SYM_OBJ_NAME \".\" #SYM_NAME \",\" #SYM_POS \"\\\"\")\n"
+                    "# define KLP_RELOC_SYMBOL(LP_OBJ_NAME, SYM_OBJ_NAME, SYM_NAME) \\\n"
+                    "   KLP_RELOC_SYMBOL_POS(LP_OBJ_NAME, SYM_OBJ_NAME, SYM_NAME, 0)\n"
+                    "#endif\n\n";
+        }
 
-    /* ... and call the method which updates the body of a function. (but that
-       is not a function!  But who cares, and Expr is a special type of Stmt
-       in clang so everything works!  */
-    fu.Update_References_To_Symbol(type->getUnderlyingExpr());
+        new_decl->print(outstr);
 
-    return VISITOR_CONTINUE;
-  }
+        if (SE.Ibt) {
+          std::string sym_mod = SE.IA.Get_Symbol_Module(sym_name);
+          if (sym_mod == "")
+            sym_mod = "vmlinux";
 
-  /* To handle symbol renames on situations like the sizeof below
-   *
-   * static char x[4];
-   *
-   * void f(void) {
-   *    char y[sizeof(x)];
-   * }
-   */
-  bool VisitArrayTypeLoc(const ArrayTypeLoc &type)
-  {
-    SymbolExternalizer::FunctionUpdater fu(SE, NewSymbolDecl, OldSymbolName, Wrap);
-    fu.Update_References_To_Symbol(type.getSizeExpr());
+          outstr << " \\\n" << "\tKLP_RELOC_SYMBOL(" << SE.PatchObject << ", " <<
+                 sym_mod << ", " << sym_name << ")";
+        }
+        outstr << ";\n";
 
-    return VISITOR_CONTINUE;
-  }
+        SE.Replace_Text(decl->getSourceRange(), outstr.str(), 1000);
 
-  private:
+        std::string replacement = "(*" + new_decl->getName().str() + ")";
```

Comment:
where is this variable used? If not, then remove it.

Reply:
This was used in the Rewrite_Macros. Let me remove it!